alicloud/resource_alicloud_snapshot.go

```diff
@@ -0,0 +1,147 @@
+	client := meta.(*connectivity.AliyunClient)
+
+	args := ecs.CreateCreateSnapshotRequest()
+	args.DiskId = d.Get("disk_id").(string)
+	if name, ok := d.GetOk("name"); ok {
+		args.SnapshotName = name.(string)
+	}
+	if description, ok := d.GetOk("description"); ok {
+		args.Description = description.(string)
+	}
+
+	raw, err := client.WithEcsClient(func(ecsClient *ecs.Client) (interface{}, error) {
+		return ecsClient.CreateSnapshot(args)
+	})
+	if err != nil {
+		return WrapErrorf(err, DefaultDebugMsg, args.DiskId, args.GetActionName(), AlibabaCloudSdkGoERROR)
+	}
+	addDebug(args.GetActionName(), raw)
+	resp := raw.(*ecs.CreateSnapshotResponse)
+	d.SetId(resp.SnapshotId)
+
+	ecsService := EcsService{client}
+	if err := ecsService.WaitForSnapshot(d.Id(), SnapshotCreatingAccomplished, DefaultLongTimeout); err != nil {
+		return fmt.Errorf("wait for snapshot (%s) creating completed failed, %s", d.Id(), err)
+	}
+
+	return resourceAliyunSnapshotUpdate(d, meta)
+}
+
+func resourceAliyunSnapshotRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+	ecsService := EcsService{client}
+	snapshot, err := ecsService.DescribeSnapshotById(d.Id())
+	if err != nil {
+		if NotFoundError(err) {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("describe snapshot (%s) failed when read, %s", d.Id(), err)
+	}
+	d.Set("name", snapshot.SnapshotName)
+	d.Set("disk_id", snapshot.SourceDiskId)
+	d.Set("description", snapshot.Description)
+
+	tags, err := ecsService.DescribeTags(d.Id(), TagResourceSnapshot)
+	if err != nil && !NotFoundError(err) {
+		return fmt.Errorf("describe tags for snapshot (%s) got error: %#v", d.Id(), err)
+	}
+	if len(tags) > 0 {
+		d.Set("tags", tagsToMap(tags))
+	}
+
+	return nil
+}
+
+func resourceAliyunSnapshotUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+
+	if err := setTags(client, TagResourceSnapshot, d); err != nil {
+		return fmt.Errorf("set tags for snapshot (%s) got error: %#v", d.Id(), err)
+	} else {
+		d.SetPartial("tags")
+	}
+
+	return resourceAliyunSnapshotRead(d, meta)
+}
+
+func resourceAliyunSnapshotDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+	ecsService := EcsService{client}
+
+	args := ecs.CreateDeleteSnapshotRequest()
+	args.SnapshotId = d.Id()
+
+	err := resource.Retry(DefaultTimeout*time.Second, func() *resource.RetryError {
+		_, err := client.WithEcsClient(func(ecsClient *ecs.Client) (interface{}, error) {
+			return ecsClient.DeleteSnapshot(args)
+		})
+		if err != nil {
+			if NotFoundError(err) {
+				return nil
+			}
+			if IsExceptedErrors(err, SnapshotInvalidOperations) {
+				return resource.RetryableError(fmt.Errorf("delete snapshot (%s) failed, %#v", d.Id(), err))
+			}
+			return resource.NonRetryableError(fmt.Errorf("delete snapshot (%s) failed, %#v", d.Id(), err))
```

Comment:
Please modify the line 128 to resource.RetryableError(err) and modify the line 130 to resource. NonRetryableError(err)